vllm/entrypoints/openai/protocol.py

```diff
@@ -283,21 +333,62 @@ def check_guided_decoding_count(cls, data):
                 "You can only use one kind of guided decoding "
                 "('guided_json', 'guided_regex' or 'guided_choice').")
         # you can only either use guided decoding or tools, not both
-        if guide_count > 1 and "tool_choice" in data and data[
-                "tool_choice"] != "none":
+        if (guide_count > 1 and "tool_choice" in data
+                and data["tool_choice"] != "none"
+                and data["tool_choice"] != "auto"):
             raise ValueError(
                 "You can only either use guided decoding or tools, not both.")
         return data
 
     @model_validator(mode="before")
     @classmethod
-    def check_tool_choice(cls, data):
-        if "tool_choice" in data and data["tool_choice"] != "none":
-            if not isinstance(data["tool_choice"], dict):
-                raise ValueError("Currently only named tools are supported.")
+    def check_tool_usage(cls, data):
+
+        if "tool_choice" in data:
+
+            # ensure that if "tool choice" is specified, tools are present
             if "tools" not in data or data["tools"] is None:
                 raise ValueError(
                     "When using `tool_choice`, `tools` must be set.")
+
+            # make sure that tool choice is either a named tool OR that it's set to "auto"
+            if data["tool_choice"] != "auto" and not isinstance(
+                    data["tool_choice"], dict):
+                raise ValueError(
+                    "`tool_choice` must either be a named tool or \"auto\". `tool_choice=\"none\" is not supported."
+                )
+
+            # ensure that if "tool_choice" is specified as an object, it matches a valid tool
+            if isinstance(data["tool_choice"], dict):
```

Comment:
Suggestion:

```python
if "tools" in data and "tool_choice" not in data:
    data["tool_choice"] = "auto"
    return data
elif not isinstance(data["tool_choice"], dict):
    return data
```

then put what was in `isinstance(data["tool_choice"], dict):` without indentation?

Reply:
Changed here - 8f6029f5

I like the structure of it this way because it's more explicitly about what condition it's checking for which i think enhances readability